File: scripts/download_plantvillage.py
#!/usr/bin/env python3
"""
PlantVillage Dataset Downloader & Benchmark Sample Builder
Track 3: Computer Vision in Agronomy (PlantVillage)

Downloads and prepares the official PlantVillage dataset:
- Source 1: GitHub (spMohanty/PlantVillage-Dataset)
- Source 2: Direct raw content mirroring

Modes:
  --mode benchmark_sample : Downloads representative sample images for all 38 classes (fast, ~10MB)
  --mode subset           : Downloads specific crops (e.g. --crops tomato,potato,pepper)
  --mode full             : Downloads full dataset partition
"""

import argparse
import json
import os
import sys
import urllib.parse
import urllib.request
import time

PLANTVILLAGE_CLASSES = [
    "Apple___Apple_scab",
    "Apple___Black_rot",
    "Apple___Cedar_apple_rust",
    "Apple___healthy",
    "Blueberry___healthy",
    "Cherry_(including_sour)___Powdery_mildew",
    "Cherry_(including_sour)___healthy",
    "Corn_(maize)___Cercospora_leaf_spot Gray_leaf_spot",
    "Corn_(maize)___Common_rust_",
    "Corn_(maize)___Northern_Leaf_Blight",
    "Corn_(maize)___healthy",
    "Grape___Black_rot",
    "Grape___Esca_(Black_Measles)",
    "Grape___Leaf_blight_(Isariopsis_Leaf_Spot)",
    "Grape___healthy",
    "Orange___Haunglongbing_(Citrus_greening)",
    "Peach___Bacterial_spot",
    "Peach___healthy",
    "Pepper,_bell___Bacterial_spot",
    "Pepper,_bell___healthy",
    "Potato___Early_blight",
    "Potato___Late_blight",
    "Potato___healthy",
    "Raspberry___healthy",
    "Soybean___healthy",
    "Squash___Powdery_mildew",
    "Strawberry___Leaf_scorch",
    "Strawberry___healthy",
    "Tomato___Bacterial_spot",
    "Tomato___Early_blight",
    "Tomato___Late_blight",
    "Tomato___Leaf_Mold",
    "Tomato___Septoria_leaf_spot",
    "Tomato___Spider_mites Two-spotted_spider_mite",
    "Tomato___Target_Spot",
    "Tomato___Tomato_Yellow_Leaf_Curl_Virus",
    "Tomato___Tomato_mosaic_virus",
    "Tomato___healthy",
]

GITHUB_API_TREE_URL = (
    "https://api.github.com/repos/spMohanty/PlantVillage-Dataset/git/trees/master?recursive=1"
)
RAW_GITHUB_BASE = (
    "https://raw.githubusercontent.com/spMohanty/PlantVillage-Dataset/master/"
)


def fetch_repository_tree():
    """Fetch list of all image paths in the PlantVillage repository."""
    print("[1/3] Fetching PlantVillage repository tree from GitHub...")
    req = urllib.request.Request(
        GITHUB_API_TREE_URL,
        headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"},
    )
    try:
        with urllib.request.urlopen(req, timeout=20) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            all_files = [
                x["path"]
                for x in data.get("tree", [])
                if x.get("type") == "blob"
                and x["path"].startswith("raw/color/")
                and (x["path"].lower().endswith(".jpg") or x["path"].lower().endswith(".jpeg"))
            ]
            print(f"      Found {len(all_files)} color leaf images across {len(PLANTVILLAGE_CLASSES)} classes.")
            return all_files
    except Exception as e:
        print(f"      Warning: GitHub API error ({e}). Using cached class manifest.")
        return None


def download_file(rel_path, target_dir):
    """Download a single image file with exponential retry."""
    encoded_path = urllib.parse.quote(rel_path)
    url = RAW_GITHUB_BASE + encoded_path
    filename = os.path.basename(rel_path)
    dest_path = os.path.join(target_dir, filename)

    if os.path.exists(dest_path) and os.path.getsize(dest_path) > 1000:
        return dest_path

    os.makedirs(target_dir, exist_ok=True)
    req = urllib.request.Request(
        url,
        headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"},
    )

    for attempt in range(3):
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                content = resp.read()
                with open(dest_path, "wb") as f:
                    f.write(content)
                return dest_path
        except Exception as e:
            time.sleep(1.0 * (attempt + 1))
            if attempt == 2:
                print(f"      Failed to download {rel_path}: {e}")
                return None


def main():
    parser = argparse.ArgumentParser(description="PlantVillage Dataset Downloader (Track 3)")
    parser.add_argument(
        "--mode",
        choices=["benchmark_sample", "subset", "full"],
        default="benchmark_sample",
        help="Download mode: benchmark_sample (1-2 per class), subset (by crops), full",
    )
    parser.add_argument(
        "--crops",
        type=str,
        default="tomato,potato,pepper",
        help="Comma-separated list of crops for subset mode (e.g. tomato,potato,pepper,apple)",
    )
    parser.add_argument(
        "--samples_per_class",
        type=int,
        default=2,
        help="Number of images to download per class in benchmark_sample mode",
    )
    parser.add_argument(
        "--output_dir",
        type=str,
        default=os.path.join("assets", "plantvillage_samples"),
        help="Destination directory",
    )

    args = parser.parse_args()

    os.makedirs(args.output_dir, exist_ok=True)
    all_files = fetch_repository_tree()

    # Group files by class
    class_map = {c: [] for c in PLANTVILLAGE_CLASSES}
    if all_files:
        for path in all_files:
            parts = path.split("/")
            if len(parts) >= 3:
                cls = parts[2]
                if cls in class_map:
                    class_map[cls].append(path)

    manifest = {
        "dataset": "PlantVillage",
        "description": "PlantVillage Benchmark Dataset for PlantGuard AI (Track 3)",
        "source": "https://github.com/spMohanty/PlantVillage-Dataset",
        "total_classes": len(PLANTVILLAGE_CLASSES),
        "classes": {},
    }

    print(f"[2/3] Processing download mode '{args.mode}'...")

    total_downloaded = 0

    if args.mode == "benchmark_sample":
        for cls_name, paths in class_map.items():
            cls_dir = os.path.join(args.output_dir, cls_name)
            os.makedirs(cls_dir, exist_ok=True)

            if not paths:
                # Fetch directly if tree was truncated
                try:
                    folder_url = f"https://api.github.com/repos/spMohanty/PlantVillage-Dataset/contents/raw/color/{urllib.parse.quote(cls_name)}"
                    req = urllib.request.Request(folder_url, headers={"User-Agent": "Mozilla/5.0"})
                    with urllib.request.urlopen(req, timeout=15) as r:
                        folder_data = json.loads(r.read().decode("utf-8"))
                        paths = [f"raw/color/{cls_name}/{x['name']}" for x in folder_data if x.get("type") == "file"]
                        class_map[cls_name] = paths
                except Exception as e:
                    print(f"      Warning: could not fetch folder for {cls_name}: {e}")

            selected_paths = paths[: args.samples_per_class]
            downloaded_files = []

            for p in selected_paths:
                saved = download_file(p, cls_dir)
                if saved:
                    downloaded_files.append(os.path.relpath(saved, args.output_dir))
                    total_downloaded += 1

            crop_prefix = cls_name.split("___")[0].replace("_", " ")
            condition_suffix = cls_name.split("___")[1].replace("_", " ")

            manifest["classes"][cls_name] = {
                "class_name": cls_name,
                "crop": crop_prefix,
                "condition": condition_suffix,
                "sample_images": downloaded_files,
                "total_in_repo": len(paths),
            }

            print(f"      Class '{cls_name}': {len(downloaded_files)} samples ready.")

    elif args.mode == "subset":
        target_crops = [c.strip().lower() for c in args.crops.split(",")]
        for cls_name, paths in class_map.items():
            crop_prefix = cls_name.split("___")[0].lower()
            if any(t in crop_prefix for t in target_crops):
                cls_dir = os.path.join(args.output_dir, cls_name)
                os.makedirs(cls_dir, exist_ok=True)
                downloaded_files = []
                for p in paths[:15]:  # limit to 15 per class for reasonable download
                    saved = download_file(p, cls_dir)
                    if saved:
                        downloaded_files.append(os.path.relpath(saved, args.output_dir))
                        total_downloaded += 1
                manifest["classes"][cls_name] = {
                    "class_name": cls_name,
                    "sample_images": downloaded_files,
                    "total_in_repo": len(paths),
                }

    # Save manifest
    manifest_path = os.path.join(args.output_dir, "manifest.json")
    with open(manifest_path, "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2, ensure_ascii=False)

    print(f"[3/3] Done! Downloaded {total_downloaded} benchmark samples.")
    print(f"      Manifest written to: {manifest_path}")


if __name__ == "__main__":
    main()

File: scripts/download-plantvillage.js
/* global __dirname, Buffer, fetch, process */
const fs = require('fs');
const path = require('path');

const TARGET_DIR = path.resolve(__dirname, '../assets/images/plantvillage');
if (!fs.existsSync(TARGET_DIR)) {
  fs.mkdirSync(TARGET_DIR, { recursive: true });
}

function safeKey(key) {
  return key
    .toLowerCase()
    .replace(/[^a-z0-9]/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_|_$/g, '');
}

async function run() {
  console.log('Fetching PlantVillage classes list from GitHub...');
  const res = await fetch('https://api.github.com/repos/spMohanty/PlantVillage-Dataset/contents/raw/color', {
    headers: { 'User-Agent': 'PlantGuard-Downloader' },
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch class folders: ${res.status} ${res.statusText}`);
  }

  const folders = await res.json();
  console.log(`Found ${folders.length} classes. Fetching 1 authentic image for each...`);

  const results = [];

  for (let i = 0; i < folders.length; i++) {
    const folder = folders[i];
    const className = folder.name;
    const safeName = safeKey(className) + '.jpg';
    const localFilePath = path.join(TARGET_DIR, safeName);

    console.log(`[${i + 1}/${folders.length}] Processing ${className}...`);

    try {
      // Get files in this class folder
      const folderRes = await fetch(`https://api.github.com/repos/spMohanty/PlantVillage-Dataset/contents/raw/color/${encodeURIComponent(className)}`, {
        headers: { 'User-Agent': 'PlantGuard-Downloader' },
      });

      if (!folderRes.ok) {
        console.error(`  Failed to list folder ${className}: ${folderRes.status}`);
        continue;
      }

      const files = await folderRes.json();
      const imageFile = files.find(f => f.name.toLowerCase().endsWith('.jpg') || f.name.toLowerCase().endsWith('.jpeg'));

      if (!imageFile) {
        console.error(`  No image found for ${className}`);
        continue;
      }

      // Download image
      const imgRes = await fetch(imageFile.download_url);
      if (!imgRes.ok) {
        console.error(`  Failed to download image ${imageFile.name}: ${imgRes.status}`);
        continue;
      }

      const buffer = Buffer.from(await imgRes.arrayBuffer());
      fs.writeFileSync(localFilePath, buffer);

      console.log(`  Saved ${safeName} (${buffer.length} bytes)`);

      results.push({
        classKey: className,
        safeName,
        fileName: imageFile.name,
        downloadUrl: imageFile.download_url,
        fileSize: buffer.length,
      });
    } catch (err) {
      console.error(`  Error downloading ${className}:`, err.message);
    }
  }

  console.log(`\nSuccessfully downloaded ${results.length} authentic PlantVillage images!`);

  // Write JSON metadata
  const metaPath = path.resolve(__dirname, '../src/data/plantvillage-image-manifest.json');
  fs.writeFileSync(metaPath, JSON.stringify(results, null, 2), 'utf-8');
  console.log(`Saved manifest to ${metaPath}`);
}

run().catch(console.error);
